Expose `reload_scripts` as `reload_open_files`

KANAjetzt opened the pull request:
Expose the `reload_scripts` method to match the current state of the 3.x branch.

As described in #83118, it is possible to replicate this method with GDScript:

```gdscript
static func reload_script(text_edit: TextEdit, source_code: String) -> void:
	var column := text_edit.get_caret_column()
	var row := text_edit.get_caret_line()
	var scroll_position_h := text_edit.get_h_scroll_bar().value
	var scroll_position_v := text_edit.get_v_scroll_bar().value

	text_edit.text = source_code
	text_edit.set_caret_column(column)
	text_edit.set_caret_line(row)
	text_edit.scroll_horizontal = scroll_position_h
	text_edit.scroll_vertical = scroll_position_v

	text_edit.tag_saved_version()
```

and set the `text_editor/behavior/files/auto_reload_scripts_on_external_change` setting to `true`.

Personally, I would prefer calling `EditorInterface.get_script_editor().reload_scripts()`, but I might not be aware of possible pitfalls caused by exposing this method.

This would also improve the porting of existing 3.5 plugins.

 
closes https://github.com/godotengine/godot-proposals/issues/8111

AThousandShips commented:
Please remove the special characters from the commit message and make it match the title of the PR

Also please open a proper proposal as asked on the issue

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3820,6 +3820,7 @@ void ScriptEditor::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_current_script"), &ScriptEditor::_get_current_script);
 	ClassDB::bind_method(D_METHOD("get_open_scripts"), &ScriptEditor::_get_open_scripts);
 	ClassDB::bind_method(D_METHOD("open_script_create_dialog", "base_name", "base_path"), &ScriptEditor::open_script_create_dialog);
+	ClassDB::bind_method(D_METHOD("reload_scripts", "p_refresh_only"), &ScriptEditor::reload_scripts, DEFVAL(false));
```
```suggestion
	ClassDB::bind_method(D_METHOD("reload_scripts", "refresh_only"), &ScriptEditor::reload_scripts, DEFVAL(false));
```

KANAjetzt replied:
*fixed [2af60be](https://github.com/godotengine/godot/pull/83267/commits/2af60be094b5e49aacaacacac177c447bffc7614)*

AThousandShips commented on doc/classes/ScriptEditor.xml:
```diff
@@ -49,6 +49,13 @@
 		</method>
+		<method name="reload_scripts">
+			<return type="void" />
+			<param index="0" name="p_refresh_only" type="bool" />
```
```suggestion
			<param index="0" name="refresh_only" type="bool" />
```

KANAjetzt replied:
*fixed [2af60be](https://github.com/godotengine/godot/pull/83267/commits/2af60be094b5e49aacaacacac177c447bffc7614)*

AThousandShips commented:
@twaritwaikar was exposing this in 3.x meant to be used in scripting, or was it renamed since it was made public on the c++ side?

Qubus0 commented:
This is useful (and used) for all plugins that modify any (active) scripts. 
I'm not sure why it would not be intended to be exposed. Would there be any downside to having it exposed or is it a case of needing a certain degree of usefulness to be justified?

AThousandShips commented:
Exposing internal methods, especially in the editor, binds the development to maintain compatibility and stability, so that's why I asked for a proposal to gauge the need and justification for this

If it can be worked around, how needed it is, etc.

Because exposing a method isn't a trivial thing that comes with possible issues and a burden of maintenance

KANAjetzt commented:
What can I do to make the Linux check pass?

AThousandShips commented:
See the error message, you need to reorder the entries of the documentation, they need to be ge sorted

KANAjetzt commented:
Sorry for that unnecessary check, I hope I got it now 😅 
*The doctool got stuck when trying to delete the cache while VSCode was open.*

Paulb23 commented:
Don't see a solid reason against exposing this for plugins, but I'm thinking we should start moving away from the `script` naming in this file a little bit as `reload_scripts` doesn't just reload `Script`, but any open file i.e `TextFile` and `JSON`.

So we should bind it as `reload_open_files` or something along those lines.

KANAjetzt commented:
Thanks, @Paulb23 - I went ahead and changed it to your suggestion. Is there something to do now on the project converter side?

KANAjetzt committed:
Expose reload_scripts as reload_open_files

Expose the reload_scripts method to match the current state of the 3.x branch.
Bind it as `reload_open_files` in contrast to `reload_scripts` in 3.x to more accurately reflect its function.

closes https://github.com/godotengine/godot-proposals/issues/8111

clarifies the description of the method's internals

Co-authored-by: Chris Cranford <ccranfor@redhat.com>

Naros commented on doc/classes/ScriptEditor.xml:
```diff
@@ -57,6 +57,13 @@
+			<return type="void" />
+			<param index="0" name="refresh_only" type="bool" default="false" />
+			<description>
+				Reload all currently opened files from disk in case the file contents are newer.
```
```suggestion
				Reload all currently opened files if [param refresh_only] is [code]false[/code]. When [param refresh_only] is [code]true[/code], only modified files since its last save are reloaded. 
				[b]Note:[/b] This method has no effect if an external editor is currently active.
```
This just clarifies the description of the method's internals, and how the argument is interpreted.

KANAjetzt replied:
great addition thank you 👍

svetogam commented:
The absence of this feature is painful to me. Two of my plugins would benefit from it. I hope it can get merged soon!

Zephilinox commented:
I also need this, my usecase is generating gdscript code in open files. the file on disk and the data in memory is updated, but the opened scripts in the editor do not get updated

Naros commented:
Correct me, but isn't this now exposed in `ScriptLanguageExtension::_reload_scripts` ?

Zephilinox commented:
I did try that before but it didn't work. I suspect it's referring to script resources, and not the open script in the script editor, or maybe it's broken or I'm using it wrong 😂

dawdle-deer commented:
> Correct me, but isn't this now exposed in `ScriptLanguageExtension::_reload_scripts` ?

`ScriptLanguageExtension` is undocumented, so I'm shooting in the dark to an extent, but I believe it's unrelated. Internally, when changes to gdscript files are saved, a call is made to `GDScriptTextDocument::reload_script`, which calls `ScriptEditor::get_singleton()->reload_scripts(true)`. It would make sense to expose this API in a similarly simple fashion.

dawdle-deer reviewed (approved):
Looks good to me!

kitbdev reviewed (commented):
This can cause unsaved changes to be lost without confirmation. 
It would be better to expose `_test_script_times_on_disk` (and maybe `_update_modified_scripts_for_external_editor` which is usually called after it?, effectively `_on_find_in_files_modified_files`) if needed.

`text_editor/behavior/files/auto_reload_scripts_on_external_change` is now true by default, so it won't show the confirmation dialog unless its unsaved.

A workaround is to send the ScriptEditor `NOTIFICATION_APPLICATION_FOCUS_IN`, since it calls `_test_script_times_on_disk`:
```gdscript
EditorInterface.get_script_editor().notification(Window.NOTIFICATION_APPLICATION_FOCUS_IN)
```
